### PR TITLE
Add ibgu package

### PR DESCRIPTION
**Pull request by mcornea**
Add ibgu package for ImageBasedGroupUpgrades custom resource.

Depends-On: https://github.com/openshift-kni/eco-goinfra/pull/663

### CODE REVIEW

**Comment by mcornea** [comment]
cc @josclark42

File: pkg/ibgu/ibgu.go [hunk 1]
```diff
@@ -0,0 +1,331 @@
...
+	return imagebasedgroupupgrade, err
+}
+
+// Exists checks whether the given imagebasedgroupupgrade exists.
+func (builder *IbguBuilder) Exists() bool {
+	if valid, _ := builder.validate(); !valid {
+		return false
+	}
+
+	glog.V(100).Infof("Checking if imagebasedupgrade %s exists",
```
Comment (achuzhoy) [inline pkg/ibgu/ibgu.go:247]:
```suggestion
	glog.V(100).Infof("Checking if imagebasedgroupupgrade %s exists",
```

File: pkg/ibgu/ibgu.go [hunk 2]
```diff
@@ -0,0 +1,331 @@
...
+			MatchLabels: labels,
+		},
+	}
+
+	builder.Definition.Spec.ClusterLabelSelectors = labelSelectors
+
+	return builder
+}
+
+// WithSeedImageRef appends the SeedImageRef to the ibuSpec.
```
Comment (josclark42) [inline pkg/ibgu/ibgu.go:109]: to the ibguSpec.

File: pkg/ibgu/ibgu.go [hunk 3]
```diff
@@ -0,0 +1,331 @@
...
+
+	if seedVersion == "" {
+		glog.V(100).Info("The 'seedVersion' parameter is empty")
+
+		builder.errorMsg = "seedVersion cannot be empty"
+
+		return builder
+	}
+
+	ibuSpec := lcav1.ImageBasedUpgradeSpec{
```
Comment (josclark42) [inline pkg/ibgu/ibgu.go:131]: ibguSpec ?

**Review by kononovn — CHANGES_REQUESTED** [review]
@mcornea Can you please split this PR into 2 different PRs. 
1 PR for sync and scheme update
2 PR for ibgu package

File: pkg/ibgu/ibgu.go [hunk 4]
```diff
@@ -0,0 +1,331 @@
...
+		SeedImageRef: lcav1.SeedImageRef{
+			Image:   seedImage,
+			Version: seedVersion,
+		},
+	}
+
+	builder.Definition.Spec.IBUSpec = ibuSpec
```
Comment (josclark42) [inline pkg/ibgu/ibgu.go:138]: builder.Definition.Spec.IBGUSpec = ibguSpec

File: pkg/ibgu/ibgu.go [hunk 5]
```diff
@@ -0,0 +1,331 @@
...
+
+	return builder
+}
+
+// WithOadpContent appends the oadpContent to the ibuSpec.
```
Comment (josclark42) [inline pkg/ibgu/ibgu.go:143]: to the ibguSpec.

File: pkg/ibgu/ibgu.go [hunk 6]
```diff
@@ -0,0 +1,331 @@
...
+	}
+
+	if builder.Object != nil {
+		err := builder.apiClient.Delete(context.TODO(), builder.Object)
+		if err != nil {
+			return err
+		}
+
+		builder.Object = nil
+	}
```
Comment (klaskosk) [inline pkg/ibgu/ibgu.go:296]: It's equivalent so not a big deal but I think using `builder.Exists()` like in [other delete functions](https://github.com/openshift-kni/eco-goinfra/blob/main/pkg/deployment/deployment.go#L430-L443) should be preferred for consistency

**Comment by mcornea** [comment]
> @mcornea Can you please split this PR into 2 different PRs. 1 PR for sync and scheme update 2 PR for ibgu package

Sure, I opened https://github.com/openshift-kni/eco-goinfra/pull/663 for the sync and schema.

File: pkg/ibgu/ibgu.go (hunk 6 above)
Comment (mcornea) [reply pkg/ibgu/ibgu.go:296]: Sure, updated. Tnx for the pointer!

File: pkg/ibgu/ibgu.go (hunk 3 above)
Comment (mcornea) [reply pkg/ibgu/ibgu.go:131]: It's actually ibuSpec as the ImageBasedGroupUpgrade spec contains an ibuSpec field based on the ImageBasedUpgradeSpec struct.

**Comment by mcornea** [comment]
I'll rebase once https://github.com/openshift-kni/eco-goinfra/pull/663 gets merged.

File: pkg/ibgu/ibgu.go [hunk 7]
```diff
@@ -0,0 +1,332 @@
...
+// WithOadpContent appends the oadpContent to the ibuSpec.
+func (builder *IbguBuilder) WithOadpContent(name string, namespace string) *IbguBuilder {
+	if valid, _ := builder.validate(); !valid {
+		return builder
+	}
+
+	if name == "" {
+		glog.V(100).Info("The 'name' parameter for OADP content is empty")
+
+		builder.errorMsg = "OADP content name cannot be empty"
```
[the remark this answers is not in the document]
Comment (mcornea) [reply pkg/ibgu/ibgu.go]: good catch, updated. Thanks!

File: pkg/ibgu/ibgu.go [hunk 8]
```diff
@@ -0,0 +1,332 @@
...
+// WithPlan appends the plan to the ibgu.
+func (builder *IbguBuilder) WithPlan(actions []string, maxConcurrency int, timeout int) *IbguBuilder {
+	if valid, _ := builder.validate(); !valid {
+		return builder
+	}
+
+	if len(actions) == 0 {
+		glog.V(100).Info("The 'actions' slice is empty")
+
+		builder.errorMsg = "Plan actions cannot be empty"
```
[the remark this answers is not in the document]
Comment (mcornea) [reply pkg/ibgu/ibgu.go]: updated

**Review by klaskosk — APPROVED** [review]
/lgtm

File: pkg/ibgu/ibgu.go [hunk 9]
```diff
@@ -0,0 +1,332 @@
...
+
+	return builder
+}
+
+// WithClusterLabelSelectors appends labels to the ibgu clusterLabelSelectors.
+func (builder *IbguBuilder) WithClusterLabelSelectors(labels map[string]string) *IbguBuilder {
+	if valid, _ := builder.validate(); !valid {
+		return builder
+	}
+
```
Comment (kononovn) [inline pkg/ibgu/ibgu.go:89]: Please add log message glog.V(100).Infof(...)

File: pkg/ibgu/ibgu.go [hunk 10]
```diff
@@ -0,0 +1,332 @@
...
+	return builder
+}
+
+// WithSeedImageRef appends the SeedImageRef to the ibuSpec.
+func (builder *IbguBuilder) WithSeedImageRef(seedImage string, seedVersion string) *IbguBuilder {
+	if valid, _ := builder.validate(); !valid {
+		return builder
+	}
+
```
Comment (kononovn) [inline pkg/ibgu/ibgu.go:114]: Please add log message glog.V(100).Infof(...)

File: pkg/ibgu/ibgu.go [hunk 11]
```diff
@@ -0,0 +1,332 @@
...
+	return builder
+}
+
+// WithOadpContent appends the oadpContent to the ibuSpec.
+func (builder *IbguBuilder) WithOadpContent(name string, namespace string) *IbguBuilder {
+	if valid, _ := builder.validate(); !valid {
+		return builder
+	}
+
```
Comment (kononovn) [inline pkg/ibgu/ibgu.go:148]: Please add log message glog.V(100).Infof(...)

File: pkg/ibgu/ibgu.go [hunk 12]
```diff
@@ -0,0 +1,332 @@
...
+	return builder
+}
+
+// WithPlan appends the plan to the ibgu.
+func (builder *IbguBuilder) WithPlan(actions []string, maxConcurrency int, timeout int) *IbguBuilder {
+	if valid, _ := builder.validate(); !valid {
+		return builder
+	}
+
```
Comment (kononovn) [inline pkg/ibgu/ibgu.go:180]: Please add log message glog.V(100).Infof(...)

**Review by kononovn — CHANGES_REQUESTED** [review]
looks good just few minor comments

File: pkg/ibgu/ibgu.go (hunk 9 above)
Comment (mcornea) [reply pkg/ibgu/ibgu.go:89]: Tnx, added.

File: pkg/ibgu/ibgu.go (hunk 12 above)
Comment (mcornea) [reply pkg/ibgu/ibgu.go:186]: Tnx, added.

File: pkg/ibgu/ibgu.go (hunk 10 above)
Comment (mcornea) [reply pkg/ibgu/ibgu.go:116]: Tnx, added.

**Pull request merged**